http-common/src/connector.rs

```diff
@@ -72,36 +68,174 @@ impl Incoming {
+            > + Clone
+            + Send
+            + 'static,
+        <H as hyper::service::Service<hyper::Request<hyper::Body>>>::Future: Send,
+    {
+        // Keep track of the number of running tasks.
+        let tasks = std::sync::atomic::AtomicIsize::new(0);
+        let tasks = std::sync::Arc::new(tasks);
+
+        let shutdown_loop = shutdown;
+        futures_util::pin_mut!(shutdown_loop);
+
+        match self {
+            Incoming::Tcp { listener } => loop {
+                let accept = listener.accept();
+                futures_util::pin_mut!(accept);
+
+                match future::select(shutdown_loop, accept).await {
+                    future::Either::Left((_, _)) => break,
+                    future::Either::Right((tcp_stream, shutdown)) => {
+                        let tcp_stream = tcp_stream?.0;
+
+                        let server = crate::uid::UidService::new(None, 0, server.clone());
+
+                        tasks.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
+                        let server_tasks = tasks.clone();
+                        tokio::spawn(async move {
+                            serve_tcp(server, tcp_stream).await;
+                            server_tasks.fetch_sub(1, std::sync::atomic::Ordering::Relaxed);
+                        });
+
+                        shutdown_loop = shutdown;
+                    }
+                }
+            },
+
+            Incoming::Unix {
+                listener,
+                user_state,
+            } => loop {
+                let accept = listener.accept();
+                futures_util::pin_mut!(accept);
+
+                // Await either the next established connection or the shutdown signal.
+                match future::select(shutdown_loop, accept).await {
+                    future::Either::Left((_, _)) => break,
+                    future::Either::Right((unix_stream, shutdown)) => {
+                        let unix_stream = unix_stream?.0;
+
+                        let (user_state, ucred) = get_user_auth(&unix_stream, user_state)?;
+                        let server =
+                            crate::uid::UidService::new(ucred.pid(), ucred.uid(), server.clone());
+
+                        tasks.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
+                        let server_tasks = tasks.clone();
+                        tokio::spawn(async move {
+                            serve_unix(user_state, server, unix_stream).await;
+                            server_tasks.fetch_sub(1, std::sync::atomic::Ordering::Relaxed);
+                        });
+
+                        shutdown_loop = shutdown;
+                    }
+                };
+            },
+        }
+
+        // Wait for all running server tasks to finish before returning.
+        let poll_ms = std::time::Duration::from_millis(100);
+
+        while tasks.load(std::sync::atomic::Ordering::Relaxed) != 0 {
```

Comment:
@arsing Is there a better way to wait for an atomic to have a certain value? Though the spin-wait isn't too bad.

Reply:
You could replace the counter with a monotonic counter + a map of current counter value -> `JoinHandle` returned by `tokio::spawn`. The counter value is the key that allows you to remove the `JoinHandle`. Then this loop is just `futures_util::future::join_all(map.into_iter().map(|(_, join_handle)| join_handle))`

On the other hand, not only isn't sleeping loop simpler, it also has the advantage that you can log "Waiting for {} more requests to complete..." before you sleep, so I think I prefer that.

Reply:
I considered keeping the handles, but you're right the counter was so much simpler. I'll keep as-is and add that log message.